libraries/auth/src/main/java/org/datatransferproject/auth/OAuth1DataGenerator.java

```diff
@@ -0,0 +1,141 @@
+    tempTokenRequest.signer = config.getRequestTokenSigner(clientSecret);
+    if (config.whenAddScopes() == OAuth1Step.REQUEST_TOKEN) {
+      tempTokenRequest.set(config.getScopeParameterName(), scope);
+    }
+    TokenSecretAuthData authData;
+    try {
+      // get request token
+      OAuthCredentialsResponse tempTokenResponse = tempTokenRequest.execute();
+      authData = new TokenSecretAuthData(tempTokenResponse.token, tempTokenResponse.tokenSecret);
+    } catch (IOException e) {
+      logger.warn("Couldn't get request token {}", e.getMessage());
+      return null;
+    }
+
+    OAuthAuthorizeTemporaryTokenUrl authorizeUrl = new OAuthAuthorizeTemporaryTokenUrl(
+        config.getAuthorizationUrl());
+    authorizeUrl.temporaryToken = authData.getToken();
+    if (config.whenAddScopes() == OAuth1Step.AUTHORIZATION) {
+      authorizeUrl.set(config.getScopeParameterName(), scope);
+    }
+    String url = authorizeUrl.build();
+
+    return new AuthFlowConfiguration(url, getTokenUrl(), AuthProtocol.OAUTH_1, authData);
+  }
+
+  @Override
+  public AuthData generateAuthData(String callbackBaseUrl, String authCode, String id,
+      AuthData initialAuthData, String extra) {
+    Preconditions
+        .checkArgument(Strings.isNullOrEmpty(extra), "Extra data not expected for OAuth flow");
+    Preconditions.checkArgument(initialAuthData != null,
+        "Initial auth data expected for " + config.getServiceName());
+
+    OAuthGetAccessToken accessTokenRequest = new OAuthGetAccessToken(config.getAccessTokenUrl());
+    accessTokenRequest.transport = httpTransport;
+    accessTokenRequest.temporaryToken = ((TokenSecretAuthData) initialAuthData).getToken();
+    accessTokenRequest.consumerKey = clientId;
+    accessTokenRequest.verifier = authCode;
+    accessTokenRequest.signer = config
+        .getAccessTokenSigner(clientSecret, ((TokenSecretAuthData) initialAuthData).getSecret());
+    TokenSecretAuthData accessToken;
+    try {
+      OAuthCredentialsResponse response = accessTokenRequest.execute();
+      accessToken = new TokenSecretAuthData(response.token, response.tokenSecret);
+    } catch (IOException e) {
+      logger.warn("Couldn't get access token {}", e.getMessage());
+      return null;
+    }
+
+    return accessToken;
+  }
+
+  private void validateConfig() {
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(config.getServiceName()),
+        "Config is missing service name");
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(config.getRequestTokenUrl()),
+        "Config is missing request token url");
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(config.getAuthorizationUrl()),
+        "Config is missing authorization url");
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(config.getAccessTokenUrl()),
+        "Config is missing access token url");
+    Preconditions
+        .checkArgument(config.getExportScopes() != null, "Config is missing export scopes");
```

Comment:
I think some APIs may not use OAuth scopes. It's probably bad in that case but technically valid OAuth1 protocol?

Reply:
Good point.  What do you think about adding a possibility for config.whenAddScopes() to return NEVER?

Reply:
Discussed offline, we're going to maintain the scopes requirement until a counterexample presents itself.

Reply:
Ack, mind just documenting here that this decision is not OAuth spec but rather us trying to encourage good practices and no accidental scope omission?